### ISSUE TITLE
Netty channel hangs after sending a payload of size > 16 MB

### ISSUE THREAD

**Issue by xiazuojie** [rsocket/rsocket-java#742]
The max allowed size of a frame is 16 MB (according to [this](https://github.com/rsocket/rsocket-java/blob/3f5b30448777b631af0de96aff9639096639a0fc/rsocket-core/src/main/java/io/rsocket/frame/FrameLengthFlyweight.java#L12)). If a payload is larger than 16 MB, encoding it to a frame will fail with an exception ["Length is larger than 24 bits"](https://github.com/rsocket/rsocket-java/blob/3f5b30448777b631af0de96aff9639096639a0fc/rsocket-core/src/main/java/io/rsocket/frame/FrameLengthFlyweight.java#L19).

However, this exception is not properly handled by [MonoSendMany](https://github.com/reactor/reactor-netty/blob/2e5158c0d09702b54f1a028b075c9ebc40b5f421/src/main/java/reactor/netty/channel/MonoSendMany.java#L272). A single such exception (encoding happens inside applying MapFuseableSubscriber#mapper) will [terminate the subscription](https://github.com/reactor/reactor-netty/blob/2e5158c0d09702b54f1a028b075c9ebc40b5f421/src/main/java/reactor/netty/channel/MonoSendMany.java#L342).  No payloads will be sent to netty channel after (due to subscription terminated).

In short, a single payload of size >= 16 MB will make the channel hang forever. And worse, there is no error log at all when this happens.

According to spec, large payloads should be [fragmented to more frames](http://rsocket.io/docs/Protocol#fragmentation-and-reassembly). However, this feature is currently not implemented transparently.

**Comment by OlegDokuka** [comment]
I will take a look at that.

Is not that a reactor-netty bug? 

cc/ @violetag

**Comment by xiazuojie** [comment]
The feature used to be in rsocket-java (SendPublisher) , later moved to reactor-netty. I think it's more about RSocket Java not properly handling payloads > 16MB.

**Comment by violetagg** [comment]
> I will take a look at that.
> 
> Is not that a reactor-netty bug?
> 
> cc/ @violetagg

Not seen so far.

**Comment by rstoyanchev** [comment]
> According to spec, large payloads should be fragmented to more frames. However, this feature is currently not implemented transparently.

There is a `fragmentation(..)` setting on `RSocketFactory` and if you set that it should fragment. If you're keeping at 0 intentionally (it's also the default) then yes it should be handled better.

**Comment by OlegDokuka** [comment]
Due to what I see, there are two issues:

1) Lack of documentation on what is the defaults in RSocket-Java and that there is no fragmentation by default. It means that all the frames that exit 16 MB size is illegal.
2) We have a bug in TCP connection which tries to add frame length at the transport level and throws an exception because the encoded frame was not validated and just simply got encoded and send (lack of Validation on RSocket level)
3) We have a bug in WebSocket which sends that frame and then the responder is not able to reassemble it back

Note, in both (2,3) we get connection termination instead of logical stream termination with the proper message

**Comment by rstoyanchev** [comment]
Given #768 to validate payload length earlier, and #769 to document, what's actually left here? 

Payloads with invalid length should not be able to get through, but arguably the transport layer should still have independent validation. Perhaps an additional check in WebSocket transports perhaps to prevent sending a frame that's too large? If that's all it is perhaps this can be added as a task under #768 and this can be closed as "superseded"?

**Comment by OlegDokuka** [comment]
@rstoyanchev 

>An additional check in WebSocket transports

This check is already present, no need to put extra one

**Comment by rstoyanchev** [comment]
I just noticed the "Linked pull requests" on the right, and in the timeline above it says this will be fixed by #768. All good.

**Comment by OlegDokuka** [comment]
@xiazuojie RC-7 should have fixed your issue. Please check it and feel free to reopen this issue if the bug was not fixes
